### PR TITLE
Adds onBeforeClose event to Modal

### PR DESCRIPTION
**Pull request by gerkirill**
### Prerequisites

Please make sure you can check the following two boxes:

- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project

### Contribution Type

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue, please reference the issue id)
- [ ] New feature (non-breaking change which adds functionality, make sure to open an associated issue first)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] My change requires a change to the documentation
- [x] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

### Description

Adds an onBeforeClose event to the modal component. The event handler can decide either the modal is allowed to close or not (depends on event handler return value). The event handler can be asynchronous. In this case, it should return a promise, which resolves to a boolean value.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -118,4 +118,4 @@
     "webpack-cli": "^3.2.1",
     "webpack-dev-server": "^3.1.14"
   }
-}
+}
```
Comment (dimabory) [inline package.json]: please return newline at the end of file

File: src/components/Modal/Example.md [hunk 2]
```diff
@@ -19,6 +19,10 @@ class ModalConsumer extends React.Component {
     e.preventDefault();
   }
 
+  handleBeforeClose() {
+  	return window.confirm('Are you sure? All unsaved data will be lost.');
+  }
+
```
Comment (dimabory) [inline src/components/Modal/Example.md]: Please use a new one example for showing how `handleBeforeClose` works.

File: src/components/Modal/Example.md [hunk 3]
```diff
@@ -287,3 +283,51 @@ class ModalConsumer extends React.Component {
 
 <ModalConsumer />
 ```
+
+**onBeforeClose**
```
Comment (dimabory) [inline src/components/Modal/Example.md]: Can you give a meaningful name?

Comment (gerkirill) [reply src/components/Modal/Example.md]: like what? 
 - "onBeforeClose event?"
 - "controlling form close"
- [your option here]

File: src/components/Modal/Example.md [hunk 4]
```diff
@@ -283,3 +283,51 @@ class ModalConsumer extends React.Component {
 
 <ModalConsumer />
 ```
+
+**Prompting on before closing**
+
+Modal with custom confirmation on close.
+
+```jsx
```
Comment (dimabory) [inline src/components/Modal/Example.md]: We can skip it as a dummy snapshot, which doesn't bring any value
```suggestion
```jsx { "props": { "data-skip": true } }
```

**Comment by dimabory** [comment]
closes #183

**Pull request merged**